leakpro/signals/utils/HopSkipJumpDIstance.py

```diff
@@ -0,0 +1,794 @@
+        -------
+            Tensor: The computed distance between the original batch and the perturbed batch.
+
+        """
+        if torch_any(torch_isnan(batch_ori)):
+            raise AssertionError("NaN values found in batch_ori")
+        assert not torch_any(torch_isnan(batch_pert)), "NaN values found in batch_ori"
+
+        if self.constraint == 2:
+            dist = norm(batch_ori - batch_pert, p=2,  dim=(1, 2, 3))
+        elif self.constraint == np.inf:
+            dist= amax(abs(batch_ori - batch_pert),  dim=(1, 2, 3))
+        return dist
+
+
+    def project(self: Self,
+                original_image_batch: Tensor,
+                perturbed_images_batch: Tensor,
+                alphas: Tensor) -> Tensor:
+        """Projects the perturbed images onto the epsilon ball around the original images.
+
+        Args:
+        ----
+            original_image_batch (Tensor): The batch of original images.
+            perturbed_images_batch (Tensor): The batch of perturbed images.
+            alphas (Tensor): The scaling factor for the perturbation.
+
+        Returns:
+        -------
+            Tensor: The projected images.
+
+        Raises:
+        ------
+            None
+
+        """
+        alphas = alphas.to(self.device)
+        alphas = alphas.view((len(original_image_batch), 1, 1, 1))
+        if self.constraint == 2:
+            projected = (1 - alphas) * original_image_batch + alphas * perturbed_images_batch
+        elif self.constraint == np.inf:
+            projected = clamp(
+                perturbed_images_batch, original_image_batch - alphas, original_image_batch + alphas
+            )
+        return projected
+
+
+    def geometric_progression_for_stepsize(self: Self,
+                                           samples: Tensor,
+                                           updates: Tensor,
+                                           dist: Tensor,
+                                           current_iteration: int,
+                                           active_indices: np.ndarray, b_i:int) -> Tensor:
+        """Calculates the step size for geometric progression in the HopSkipJumpDistance algorithm.
+
+        Args:
+        ----
+            samples (Tensor): The input samples.
+            updates (Tensor): The updates to be applied to the samples.
+            dist (Tensor): The distance value.
+            current_iteration (int): The current iteration of the algorithm.
+            active_indices (np.ndarray): The active indices.
+            b_i (int): The value of b_i.
+
+        Returns:
+        -------
+            Tensor: The calculated step size for geometric progression.
+
+        """
+        batch_epsilon = dist / np.sqrt(current_iteration)
+        active_indices = np.array(active_indices)
+
+        success = np.zeros(len(samples), dtype=bool)
+        batch_active_indices = np.arange(len(samples))
+        num_evals = 0
+
+        while not success.all() and num_evals <= 10:
```

Comment:
num_evals is hard coded.

Reply:
I think it makes sense to have a num_evals here. It restricts the number of queries we can make to the API. This is a quantity I think the user should be able to set!